Update effects.h to handle return calls correctly

tlively opened the pull request:
As far as their surrounding code is concerned return calls are no different from
normal returns. It's only from a caller's perspective that a function containing
a return call also has the effects of the return-callee. To model this more
precisely in EffectAnalyzer, stash the effects of return-callees on the side and
only merge them in at the end when analyzing the effects of a full function
body.

tlively commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6470?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#6470** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6470?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#6464** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6464?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6451** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6451?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6448** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6448?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @tlively and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

tlively commented:
@kripken, any suggestions you have about how to easily test this would be much appreciated.

kripken commented:
How does a non-exception effect get noticed here? That is, when we discussed this before we talked about stashing a thrown exception on the side, and not all the effects. This does more work to stash them all, and I'm not sure why yet. (I get that technically the call executes after the return, so effects happen "later" - but it seems simpler and sufficient to track only exceptions IIANM.)

For testing throwing, you need global effects, see `test/lit/passes/global-effects.wast`. E.g. the call to `$nop` is optimized away thanks to global effects there. For our purposes here you can do a call to a function whose body is a try-catch-all of a return_call, so the only effect is a throw. By not optimizing it away we prove we still see that effect. And  for comparison if the return_call were a call we do optimize it away.

tlively commented:
My hope is that more precisely analyzing return call effects would lead to more optimization opportunities, but maybe that's not the case if control flow transfers cannot be reordered with any other side effect?

tlively commented:
Ok, I implemented the simplified version that only stashes the `throws` effect in [Simplify to only stash the throws effect](https://github.com/WebAssembly/binaryen/pull/6470/commits/e88a86a0f3e3fb836043541e43ea8c400990f83e). But intentionally making the analysis less precise didn't sit right with me, so then I implemented a nicer design for stashing all the effects in [Stash all effects inline in the effect analyzer](https://github.com/WebAssembly/binaryen/pull/6470/commits/40f26ee54fd65817a7b8830c806bd949c321b24c).

Even if it doesn't make a functional difference now, being more precise by stashing all the effects can only possibly produce better results in the face of future optimizations for which it might make a difference. It also means we won't have to redo any of this work in the future if/when we add separate `suspends` or `accessesDynamicScope` or other potential scope-based effects.

kripken commented:
I see your point that stashing effects could in theory help in the future. But do you see a practical way that could end up happening? That's what I am finding hard to see.

From my perspective, a return call is like a call except with a slight reordering, which has the quirk of causing the call to happen outside of current scoping rules. But that is all. And we do not track a CFG here, so tracking where the branch goes to is beyond the scope of this class - for that reason I am skeptical it will ever benefit from the theoretical additional precision.

tlively commented:
Yeah, I guess I'm convinced that stashing will only ever be useful for effects that interact with scopes like throwing. Ok, back to the single `bool`.

kripken commented on src/ir/effects.h:
```diff
@@ -466,43 +481,61 @@ class EffectAnalyzer {
+        // same as the code below for the general path. We can always filter out
+        // throws for return calls because they are already more precisely
+        // captured by `hasReturnCallThrow`.
+        if (targetEffects->throws_ && (parent.tryDepth > 0 || curr->isReturn)) {
```
The `isReturn` part here is not clear to me, because `hasReturnCallThrow` only takes effect in `walk(Function*)`. But most users of this class do not use that, that is, they do not walk an entire function. For them, this would be removing the throws effect here - isn't that dangerous?

Concretely, Vacuum will remove this dropped call, I worry:
```wat
(func $caller
  (return_call $throw)
)
(func $throw
  (throw ..)
)
```
I worry that is the case because Vacuum computes effects for each expression it sees, including that `return_call`. The `return_call` happens to be the entire function body, but Vacuum is just walking all the expressions and doesn't know that, so it calls `walk(Expression*)` and not `walk(Function*)`.

That is, I think it is unsafe to leave things here for function-level handling because we are not always looking at the entire function.

tlively replied:
I hope that the `branchesOut` effect will prevent that kind of problem. Do you think it will?

kripken replied:
I suppose `branchesOut` fixes that specific case, good point. But if we had a testcase that was sensitive to `throws` specifically we could be in trouble. I admit I can't think of one, but I think the principle remains that we should not dpeendepend on `hasReturnCallThrow` to save us because that only runs at the function level by design, and we are not always scanning at that level.

kripken replied:
Yes, I think that's right @vouillon - after thinking more on this I agree. Overall I'd explain the rationale like this:

1. `return_call` is basically like syntax sugar for a branch to an outer block at the function's exit where a call occurs. This is the key thing to keep in mind.
2. The call will definitely execute, so it's fine to just apply its effects here, normally.
3. However, a noticeable difference is when the call's effects can be removed due to scoping, which atm includes only try-catch (but may include more in future specs). In that case we do need to model the effect as not occurring here but rather as occurring in the outer block.
4. To do that, we can remove `throws_` and keep the normal `branchesOut`, which exactly represents a branch to another place, and that place is the outer scope that has the `throws_`. Users of the code that just consider a single expression will call `walk(Expression*)` and they cannot remove `branchesOut` effects because it is a branch to some place that can have arbitrary effects (even more powerful than `throws_`, and which they don't see), so this is safe.
5. And, in `walk(Function*)` we apply that outer block and the effects of the calls that occur there, that is, users of the code that want to consider entire functions must use that form. It would be a bug for them not to, so that is something users must be careful about.

So I think this PR is correct as it is but perhaps @tlively we can add that rationale or something like it?

tlively replied:
Most of this was already captured in the comments, but I just pushed a commit adding extra detail. Let me know if you have suggestions for further clarifications.

tlively commented on test/lit/passes/global-effects.wast:
```diff
@@ -219,6 +219,53 @@
+  ;; INCLUDE:      (func $return-call-throw-and-catch (type $0)
+  ;; INCLUDE-NEXT:  (return_call $throw)
+  ;; INCLUDE-NEXT: )
+  (func $return-call-throw-and-catch
+    (try
+      (do
+        ;; This call cannot be optimized out, as the target throws. However, the
+        ;; surrounding try-catch can be removed even without global effects
+        ;; because the throw from the return_call is never observed by this
+        ;; try-catch.
+        (return_call $throw)
+      )
+      (catch_all)
+    )
+  )
```
@kripken, the part where we do not include the `throws` effect on the return_call itself is important for this test case. If we treat the `return_call` as throwing, we cannot optimize away the surrounding try-catch.

kripken replied:
Makes sense. But for correctness I think we need to fix the issue I mentioned.

I admit it seems theoretical atm, but e.g. `ignoreControlFlowTransfers()` will remove `branchesOut` which you mentioned as saving us from misoptimizing in that case. Now, `ignoreControlFlowTransfers()` is only used in one place that looks at more than a single instruction, so it ends up ok (it's not looking at the entire function), but this seems brittle.

tlively replied:
How would you fix the issue without regressing reasonable optimizations?

kripken replied:
I'm not sure yet, to be honest - there is definitely something awkward here in how this class doesn't quite match up to what we need it to do, given how `return_call` actually works. I'll think about this some more.

tlively commented:
Fuzzer seems happy with this. 20k+ iterations passed (without the last commit, which I'm fuzzing now).

vouillon commented:
Two other possible tests. It seems the code folding pass could be improved to deal better with tail calls. I had to add a `return` to make it work.
```wasm
 (func $code-folding (param $x i32)
  (try
   (do
    (drop (i32.const 1))
    (drop (i32.const 1))
    (return_call $throw)
    (return)
   )
   (catch_all
    (drop (i32.const 1))
    (drop (i32.const 1))
    (return_call $throw)
    (return)
   )
  )
 )

 (func $simplify-locals (param $x i32)
  (local $y i32)
  (local.set $y
   (if (result i32)
    (local.get $x)
    (then
     (local.get $x)
    )
    (else
     (return_call $throw)
    )
   )
  )
  (try
   (do
    (drop
     (local.get $y)
    )
   )
   (catch_all)
  )
 )
```

tlively commented:
Thanks for the suggestions, @vouillon!

kripken commented on src/ir/effects.h:
```diff
@@ -66,10 +66,16 @@ class EffectAnalyzer {
   // noticeable from the perspective of the caller, that is, effects that are
   // only noticeable during the call, but "vanish" when the call stack is
-  // unwound.
+  // unwound. Unlike walking just the body, walking the function will also
+  // include the effects of any return calls the function makes.
```
```suggestion
  // unwound.
  //
  // Unlike walking just the body, walking the function will also
```suggestion
  // include the effects of any return calls the function makes. For that
  // reason, it is a bug if a user of this code calls walk(Expression* and not
  // walk(Function*) if their intention is to scan an entire function body.
  // Putting it another way, a return_call is syntax sugar for a return and a
  // call, where the call executes at the function scope, so there is a
  // meaningful difference between scanning an expression and scanning
  // the entire function body.
```

kripken commented on src/ir/effects.h:
```diff
@@ -143,6 +149,18 @@ class EffectAnalyzer {
+  // body.
+  //
+  // We currently do this stashing only for the throw effect, but in principle
+  // we could do it for all effects if it made a difference.
```
```suggestion
  // we could do it for all effects if it made a difference. (Only throw is noticeable
  // now because the only thing that can change between doing the call here
  // and doing it outside at the function exit is the scoping of try-catch blocks.
  // If future wasm scoping additions are added, we may need more here.)
```

kripken commented on src/ir/effects.h:
```diff
@@ -466,43 +484,61 @@ class EffectAnalyzer {
+        // target function throws and we know that will be caught anyhow, the
+        // same as the code below for the general path. We can always filter out
+        // throws for return calls because they are already more precisely
+        // captured by `hasReturnCallThrow`.
```
```suggestion
        // captured by `branchesOut`. That is, return calls add`branchesOut` which
        // represents the fact that we are branching to another place that may
        // have more effects, which can include throwing or anything else that
        // happens when the call executes.
```

kripken commented on src/ir/effects.h:
```diff
@@ -466,43 +484,61 @@ class EffectAnalyzer {
-      if (parent.features.hasExceptionHandling() && parent.tryDepth == 0) {
+      // When EH is enabled, any call can throw. Skip this for return calls
+      // because the throw is already more precisely captured by
+      // `hasReturnCallThrow`.
```
```suggestion
      // `branchesOut`.
```

tlively replied:
`branchesOut` isn't related to the throw effect, though. It's the fact that we set `hasReturnCallThrow` that ensures the throw is eventually reported correctly.

kripken replied:
Correct, but the context here is removing the throws effect on the next line. It is only valid to remove the throws effect because `branchesOut` indicates "there is a branch from here to some other effects". Those other effects are `hasReturnCallThrow`, but those do not come into play normally - not unless you walk the entire function. So it's the combination of the two, but `branchesOut` is the more immediate thing that "replaces" the throws we are removing, it kind of points to where the effects are at.

tlively replied:
Consider that we still apply `branchesOut` even if the callee does not throw, though.

I'll update the comment to mention both.

kripken replied:
We do, yes. But in the `return_call` case that has the additional important meaning of "linking" us to the effects we are removing here. So it explains why the removal of `throws` is valid in that case. That is, `branchesOut` becomes even more significant for return calls given this optimization here, and if we some day refactor `branchesOut` we'd need to consider this place.

kripken commented on test/lit/passes/global-effects.wast:
```diff
@@ -219,6 +219,53 @@
+      (catch_all)
+    )
+    ;; This cannot be optimized out at all.
+    (call $return-call-throw-and-catch)
```
Please test `return_call_indirect/ref` as well, so we have coverage for them adding the `hasReturnCallThrow` field. (IIANM the tests below test the other property, that they do not have `throws_` set and so they can be optimized better.)

tlively replied:
Added tests, but I'm not sure they test what you're looking for because the callees also have the `calls` effect, which is enough to inhibit optimizations on its own.

kripken replied:
Yeah, I guess `calls` masks this. Still, those tests look useful - maybe some day we'll make `calls` more refined somehow, and these would become more important.

kripken reviewed (approved):
lgtm % comments
